app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/ConfirmResponsibleIndividualVerifiedController.scala

```diff
@@ -0,0 +1,238 @@
+          _                 <- fromOptionF(submissionReviewService.updateVerifiedByDetails(verifiedByDetails)(request.submission.id, request.submission.latestInstance.index), log("Failed to find existing review"))
+          _                 <- fromOptionF(submissionReviewService.updateActionStatus(SubmissionReview.Action.ConfirmResponsibleIndividualVerified, SubmissionReview.Status.Completed)(request.submission.id, request.submission.latestInstance.index), log("Failed to find existing review"))
+        } yield success(verifiedByDetails)
+      ).fold(identity(_), identity(_))
+    }
+
+    def comeBackLater() = {
+      (
+        for {
+          _                 <- fromOptionF(submissionReviewService.updateActionStatus(SubmissionReview.Action.ConfirmResponsibleIndividualVerified, SubmissionReview.Status.InProgress)(request.submission.id, request.submission.latestInstance.index), log("Failed to find existing review"))
+        } yield checklist
+      ).fold(identity(_), identity(_))
+    }
+
+    val formValues = request.body.asFormUrlEncoded.getOrElse(Map.empty).filterNot(_._1 == "csrfToken")
+    val submitAction = formValues.get("submit-action").flatMap(_.headOption)
+
+    submitAction match {
+      case Some("checked")         => saveAndContinue
+      case Some("come-back-later") => comeBackLater
+      case _                       => successful(log("No submit-action"))
+    }
+  }
+
+  private def getVerifiedByDetails(verified: Boolean, dateVerifiedDay: String, dateVerifiedMonth: String, dateVerifiedYear: String): Option[SubmissionReview.VerifiedByDetails] = {
+    val dateAsString = dateVerifiedYear + "-" + dateVerifiedMonth + "-" + dateVerifiedDay + "T00:00:00.000Z"
+    try {
+      val timestamp: DateTime = DateTime.parse(dateAsString)
+      Some(SubmissionReview.VerifiedByDetails(verified, Some(timestamp)))
+    } catch {
+      case e: Exception => None
+    }
+  }
+
+  def action2(applicationId: ApplicationId): Action[AnyContent] = loggedInWithApplicationAndSubmission(applicationId) { implicit request =>
+    val log = logBadRequest(SubmissionReview.Action.ConfirmResponsibleIndividualVerified) _
+    val checklist = Redirect(uk.gov.hmrc.apigatekeeperapprovalsfrontend.controllers.routes.ChecklistController.checklistPage(applicationId))
+    val failed = (msg: String, day: String, month: String, year: String) => {
+      BadRequest(confirmResponsibleIndividualVerified2Page(ViewModel2(request.application.name, applicationId, Some(day), Some(month), Some(year), Some(msg))))
+    }
+    
+    def saveAndContinue() = {
+      (
+        for {
+          submissionReview  <- fromOptionF(submissionReviewService.findReview(request.submission.id, request.submission.latestInstance.index), log("No submission review found"))
+          dateVerifiedDay   <- fromOption(request.body.asFormUrlEncoded.getOrElse(Map.empty).get("date-verified-day").flatMap(_.headOption), log("No date-verified-day found in request"))
```

Comment:
Instead of pulling these out of the request by hand, is it worth create a form for the values and just using that?
Like we do in other places -> https://github.com/hmrc/api-gatekeeper-approvals-frontend/blob/main/app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/DeclinedJourneyController.scala#L41

Reply:
Fixed